anchore_engine/clients/grype_wrapper.py

```diff
@@ -16,442 +16,467 @@
-    )
-    latest_grype_db_file = os.path.join(latest_grype_db_dir, VULNERABILITY_FILE_NAME)
-    db_connect = "sqlite:///{}".format(latest_grype_db_file)
-    latest_grype_db_engine = sqlalchemy.create_engine(db_connect, echo=True)
-    return latest_grype_db_engine
-
-
-def _init_latest_grype_db_session(grype_db_engine):
-    """
-    Create and return the db session
-    """
-    logger.info(
-        "Creating new grype_db session from engine based on {}".format(
-            grype_db_engine.url
-        )
-    )
-    SessionMaker = sessionmaker(bind=grype_db_engine)
-    grype_db_session = SessionMaker()
-    return grype_db_session
-
-
-def _init_latest_grype_db(lastest_grype_db_archive: str, version_name: str):
-    """
-    Write the db string to file, create the engine, and create the session
-    Return the file and session
-    """
-    latest_grype_db_dir = _move_and_open_grype_db_archive(
-        lastest_grype_db_archive, version_name
-    )
-    latest_grype_db_engine = _init_latest_grype_db_engine(latest_grype_db_dir)
-    latest_grype_db_session = _init_latest_grype_db_session(latest_grype_db_engine)
-
-    # Return the dir, file, and engine
-    return latest_grype_db_dir, latest_grype_db_session
-
-
-def _remove_local_grype_db(grype_db_dir):
-    """
-    Remove old the local grype db file
-    """
-    if os.path.exists(grype_db_dir):
-        logger.info("Removing old grype_db at {}".format(grype_db_dir))
-        shutil.rmtree(grype_db_dir)
-    else:
-        logger.error(
-            "Failed to remove grype db at {} as it cannot be found.".format(
-                grype_db_dir
+    def __new__(cls):
+        if cls._grype_wrapper_instance is None:
+            cls._grype_wrapper_instance = super(GrypeWrapperSingleton, cls).__new__(cls)
+            cls._grype_db_dir_internal = None
+            cls._grype_db_session_internal = None
+            cls._grype_db_lock = rwlock.RWLockWrite()
+        return cls._grype_wrapper_instance
+
+    @classmethod
+    def get_instance(self):
+        """
+        Returns the singleton instance of this class.
+        """
+        return GrypeWrapperSingleton()
+
+    @classmethod
+    def _get_test_instance(self):
+        """
+        Returns a new instance of this class. This method is not intended for use outside of tests.
+        """
+        self._grype_wrapper_instance = None
+        return GrypeWrapperSingleton()
+
+    @property
+    def _grype_db_dir(self):
+        if self._grype_db_dir_internal is None:
+            raise ValueError(self.MISSING_GRYPE_DB_DIR_ERROR_MESSAGE)
+        else:
+            return self._grype_db_dir_internal
+
+    @_grype_db_dir.setter
+    def _grype_db_dir(self, _grype_db_dir_internal):
+        self._grype_db_dir_internal = _grype_db_dir_internal
+
+    @property
+    def _grype_db_session(self):
+        if self._grype_db_session_internal is None:
+            raise ValueError(self.MISSING_GRYPE_DB_SESSION_ERROR_MESSAGE)
```

Comment:
Another pattern you might consider is using the None type check to initialize the variable rather than raising a ValueError. So you would potentially call _move_and_open_grype_db_archive, but hiding that behind a getter might be more confusing than not.

Reply:
I'd prefer to keep this a dumb client - the caller is responsible for the logic for kicking off the reinitialization og grype_db if this is called before those variables are set.